libgalois/include/katana/PropertyGraph.h

```diff
@@ -135,6 +144,49 @@ class KATANA_EXPORT GraphTopology {
 
   bool empty() const noexcept { return num_nodes() == 0; }
 
+  /**
+   * @param node node to get degree for
+   * @returns Degree of node N
+   */
+  size_t degree(Node node) const noexcept {
+    auto edge_range = edges(node);
+    return std::distance(edge_range.begin(), edge_range.end());
+  }
+
+  Edge original_edge_id(const Edge& eid) const noexcept { return eid; }
+
+  Node original_node_id(const Node& nid) const noexcept { return nid; }
```

Comment:
These need to be documented. "Original" could mean a lot of things.

Reply:
Yes, I agree. My other choice was to introduce "PropertyIndex", but then I realized that these are in fact NodeIDs/EdgeIDs, except that they are original, unsorted NodeIDs.

Reply:
We can go two ways here (adding @scober for comments too as he mentioned property_index):
1) We introduce PropertyIndex concept and these methods become `node_property_index(Node)` and `edge_property_index(Edge)` where the node/edge ID passed is the renumbered ID (and not the original ID). All properties are accessed through this property_index. (Note that this will change how `GetData/GetEdgeData` works for TypedPropertyGraph). Also, say that we implement grouping of Properties, you can imagine that we'll keep some map from Node IDs to property index. Now, all topologies will keep their own map for their renumbered Node/EdgeIDs to PropertyIndex, as opposed to a central map in option 2. 
 
2) We establish the invariant that properties can always be accessed with just Node/Edge ID and optional topologies have the methods above to obtain the unsorted Node/Edge IDs.

Reply:
I like option 1 for the understandability reasons mentioned by Roshan and Arthur.

I would like to push that suggestion to include the base `PropertyGraph`, though. So that even if you have a "regular" `PropertyGraph` and not a view, you still need to call `{node,edge}_property_index`.

That would bring the interfaces closer together and give the sense that the base `PropertyGraph` is still a view, it is just the default view. I also like the idea that the partitioner could create property arrays + an indirection layer for the base `PropertyGraph` instead of a property table ordered by node/edge ids.

Reply:
I don't fully understand option 2 here. But I think option 1 sounds good. And it must be used in both the views and the main PropertyGraph for consistency.

Reply:
I'm in favor of (1). That fits well into our long term plans.

Reply:
Hi guys, I know we all like option (1) but I'm a bit hesitant. Here's why. With option (1), every topology maintains its own map from Node/Edge IDs to PropertyIndex. Additionally, topologies that are Label/Type aware will need to copy the mapping from Node/Edge IDs to TypeSetIDs and rearrange it upon sorting, just like the PropertyIndex. 

Now, with option (2), we don't copy any of these mappings. There's only mapping from original (unsorted) Node/Edge IDs to PropertyIndex, and similarly for TypeSetIDs. The optional topologies keep a mapping from sorted/rearranged Node/Edge IDs to original Node/Edge IDs. So a PropertyIndex (or TypeSetID) lookup goes as sorted Node ID ---> unsorted ID ---> PropertyIndex.  In a way, option(1) is flattening this level of indirection to sorted Node ID ---> PropertyIndex, at the cost of extra memory (and extra swaps). 

I'm unsure at the moment and tempted to leave it as is until we implement grouping of properties and have non-trivial PropertyIndices. Thoughts?

Reply:
Thanks. I didn't understand the extra memory overhead. Why would the overhead be different? Every topology, except original, will need a map from node ID -> property index or unsorted node ID, right?

Also, original topology will also need mapping from node ID -> property index once we compact properties. If the mapping is identity, then we can specialize for that.

Reply:
I think we are mixing API and implementation here. I'm fine with having two levels of indirection. That's not problem. The important this is that in the API we do not require that the user does the mapping themselves. And I like the API implied by (1). So if we can implement that API using the more compact set of tables, I would we perfectly happy.

Reply:
OK, I hear you guys and I agree, but I'm going to postpone this change to my next PR that will implement the PropertyGraph Views. I'd like to close this PR without introducing more changes (apart from addressing the comments). 

@roshandathathri : there is a minor difference in overheads between options (1) and (2), and it may not matter. With (1), we need to copy typeID arrays and sort them, but on the plus side, we remove one level of indirection in look ups. With (2), we only need to maintain sorted ID --> unsorted ID mapping, so we sort only the ID array (in addition to destinations array). 

There is one point to ponder. Having topologies maintain property_index for lookups means that we need to define the interface between topologies and the (non-existent as of now) Property Store. Property Store is what will remain of PropetyGraph once we remove the topology related guts from it and leave only Property related code. We will still have PropertyGraph which will be the view with default (original/unsorted) topology.

Reply:
OK. I get that. Thanks for cutting off the bike shedding.

My specific request is to document the API **in this PR** so it's really clear which type of ID is expected by and returned by each method. Is it an original/default ID? Is it a ID for the sorted view? For another view? That way we can a) correctly use this API while it exists and b) have a good way to talk about the various types of IDs while we hide them behind abstractions.